lib/types.ts

```diff
@@ -265,44 +265,61 @@ type Collection<TKey extends CollectionKeyBase, TMap, TValue> = {
 };
 
 type WithOnyxInstance = Component<unknown, WithOnyxInstanceState<NullableKeyValueMapping>> & {
-    setStateProxy: (cb: (state: NullableKeyValueMapping) => OnyxValue<OnyxKey>) => void;
+    setStateProxy: (cb: (state: Record<string, OnyxCollection<KeyValueMapping[OnyxKey]>>) => OnyxValue<OnyxKey>) => void;
     setWithOnyxState: (statePropertyName: OnyxKey, value: OnyxValue<OnyxKey>) => void;
 };
 
 /** Represents the base options used in `Onyx.connect()` method. */
 type BaseConnectOptions = {
-    statePropertyName?: string;
-    withOnyxInstance?: Component;
     initWithStoredValues?: boolean;
 };
 
+/** Represents additional options used inside withOnyx HOC */
+type WithOnyxConnectOptions<TKey extends OnyxKey> = {
+    withOnyxInstance: WithOnyxInstance;
+    statePropertyName: string;
+    displayName: string;
+    initWithStoredValues?: boolean;
+    selector?: Selector<TKey, unknown, unknown>;
+    canEvict?: boolean;
+};
+
+type DefaultConnectCallback<TKey extends OnyxKey> = (value: OnyxEntry<KeyValueMapping[TKey]>, key: TKey) => void;
+
+type CollectionConnectCallback<TKey extends OnyxKey> = (value: OnyxCollection<KeyValueMapping[TKey]>) => void;
+
+/** Represents the callback function used in `Onyx.connect()` method with a regular key. */
+type DefaultConnectOptions<TKey extends OnyxKey> = {
+    key: TKey;
+    callback?: DefaultConnectCallback<TKey>;
+    waitForCollectionCallback?: false;
+};
+
+/** Represents the callback function used in `Onyx.connect()` method with a collection key. */
+type CollectionConnectOptions<TKey extends OnyxKey> = {
+    key: TKey extends CollectionKeyBase ? TKey : never;
+    callback?: CollectionConnectCallback<TKey>;
+    waitForCollectionCallback: true;
+};
```

Comment:
Last comment/question: Wondering why these types were not defined like this:
```suggestion
type DefaultConnectOptions<TKey extends Key> = {
    key: TKey;
    callback?: DefaultConnectCallback<TKey>;
    waitForCollectionCallback?: false;
};

/** Represents the callback function used in `Onyx.connect()` method with a collection key. */
type CollectionConnectOptions<TKey extends CollectionKeyBase> = {
    key: TKey;
    callback?: CollectionConnectCallback<TKey>;
    waitForCollectionCallback: true;
};
```

`DefaultConnectOptions<TKey extends OnyxKey>` feels a bit wrong since `OnyxKey` can be a collection key, then the `waitForCollectionCallback: false` would be wrong for such key. Also `CollectionConnectOptions<TKey extends CollectionKeyBase> ` allows you to do `key: TKey` instead of the more complex `TKey extends CollectionKeyBase ? TKey : never;`.

Do you see any downside of specifying these types like this?

Reply:
I was like, the type looks weird indeed but it got me thinking that it was defined like this for a reason 😅 

I checked in NewDot, this type works until you use it with a key (not a collection key) and add `waitForCollectionCallback: true`:
```ts
let allCards: OnyxValues[typeof ONYXKEYS.CARD_LIST] = {};
Onyx.connect({
    key: ONYXKEYS.CARD_LIST,
    callback: (val) => {
        if (!val || Object.keys(val).length === 0) {
            return;
        }

        allCards = val;
    },
    waitForCollectionCallback: true,
});
```

It makes `val: OnyxCollection<CardList>` while it should be `val: OnyxEntry<CardList>`. With the old type errors properly: `Type 'true' is not assignable to type 'false'`

Reply:
Oh, interesting, wouldn't that mean that `ONYXKEYS.CARD_LIST` is a valid `CollectionKeyBase` when it shouldn't be?

Reply:
In you example, if things work properly, I would expect two problems:

- The callback takes as param `val: OnyxCollection<CardList>`, and
- `key: ONYXKEYS.CARD_LIST` is wrong because `key` must be a `CollectionKeyBase` and is not

Reply:
> Oh, interesting, wouldn't that mean that ONYXKEYS.CARD_LIST is a valid CollectionKeyBase when it shouldn't be?

I checked it in ONYXKEYS.ts in NewDot and `ONYXKEYS.CARD_LIST` is not a valid `CollectionKeyBase`:
```ts
const cardList1: OnyxCollectionKey = 'cardList'; // errors
const cardList2: OnyxValueKey = 'cardList'; // works fine
```

For context, `OnyxCollectionKey` and `OnyxValueKey` are used in `src/types/modules/react-native-onyx.d.ts` to augment onyx keys:
```ts
declare module 'react-native-onyx' {
    // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
    interface CustomTypeOptions {
        keys: OnyxValueKey | OnyxFormKey | OnyxFormDraftKey;
        collectionKeys: OnyxCollectionKey;
        values: OnyxValues;
    }
}
```

This way keys, and collectionKeys are no longer `string` but instead are union of [literal strings](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types)

Reply:
> In your example, if things worked properly, I would have expected one of these two errors:
>
> key: ONYXKEYS.CARD_LIST is wrong because key must be a CollectionKeyBase and is not, or
waitForCollectionCallback: true has an error because ONYXKEYS.CARD_LIST is a Key and therefore, waitForCollectionCallback must be false or undefined

As you already noted, all `Key`, `OnyxKey` and `CollectionKeyBase` are of type `string` in react-native-onyx. It's later augmented in the end-user repo like NewDot, that's why it is so tricky to understand. 

While [this](https://github.com/Expensify/react-native-onyx/pull/523#discussion_r1566273457) seems like a great idea, it comes with a problem. It does work until all `Key`, `OnyxKey` and `CollectionKeyBase` are of type `string`, but once we augment these values through `src/types/modules/react-native-onyx.d.ts` file, `ConnectOptions` type breaks:
```ts
type Key = 'account' | 'isSidebarLoaded' ;
type CollectionKeyBase = 'report_';
type OnyxKey = Key | CollectionKey;

type DefaultConnectOptions<TKey extends Key> = {
    key: TKey;
    callback?: DefaultConnectCallback<TKey>;
    waitForCollectionCallback?: false;
};

type CollectionConnectOptions<TKey extends CollectionKeyBase> = {
    key: TKey;
    callback?: CollectionConnectCallback<TKey>;
    waitForCollectionCallback?: true;
};

type ConnectOptions<TKey extends OnyxKey> = CollectionConnectOptions<TKey> | DefaultConnectOptions<TKey>; 
// Type 'TKey' does not satisfy the constraint 'Key'.
  // Type 'OnyxKey' is not assignable to type 'Key'.
``` 

That is why TS struggles to show an error [here](https://github.com/Expensify/react-native-onyx/pull/523#discussion_r1566360988).

Reply:
I think we should leave `ConnectOptions` as is, @fabioh8010 maybe you have some idea as you created this type?

Reply:
I see, well, thanks a lot for trying and following up with this. Meanwhile, I'll merge since the PR leaves us in a better place anyway and I don't want to keep blocking here. If you come up with some improvements later, just lets do them in a follow up PR.

Reply:
This is the first time I see type "augmentation" in typescript, pretty cool feature!

Reply:
I think is better to keep the way it is for now, this was the best solution we found at that time to add this type safety around `waitForCollectionCallback` and collection keys.